### DISCUSSION TITLE
cloudinary

### DISCUSSION THREAD

**Discussion by DorisW3** [neuefische/questions#706] [Web] [answered]
Hey folks,

I have trouble inserting the Cloudinary setup into my project. I created an api/images/index.js folder structur and checked the route with localhost, but there is always the 404 error...

Has someone a hint, of what the problem could be?

I checked the discussion board for more information but found nothing. However, I checked the further steps and did all of them correctly.

Thank you.

**Comment by Kheirah** [comment]
Can't inspect your code if you don't make it available 😄 
Can you push your changes and create a PR?

**Comment by DorisW3** [comment, marked as answer]
This problem has already solved. The _api/images/index.js_ structure has to be in the _pages_ folder.
But I have a new error which I can't solve:
![Bildschirmfoto 2023-08-30 um 10 27 05](https://github.com/neuefische/questions/assets/135646413/5329841b-5aa5-4e7d-9bcc-9659c6a065c0)

Here one can see my PR:
https://github.com/DorisW3/capstone-project/pull/15

**Reply by Kheirah** [reply]
Hmm, can't speak to the exact object that's returned, but you can try adding _optional chaining_ and see if it helps.
```js
data.resources?.map
```
If this doesn't help, inspect the `data` object and check if it contains a "resources" attribute. Maybe it's spelled differently, maybe it's called something else entirely.

**Comment by DorisW3** [comment]
Thank you, that helped. The error has gone. :-)

**Comment by DorisW3** [comment]
Hi,

now there is an other problem, that must be solved...
I get the following error with the status of "loading..." in the image upload form, when I try to upload an Image:

![Bildschirmfoto 2023-08-30 um 12 13 52](https://github.com/neuefische/questions/assets/135646413/67b952e6-2a8f-4b8d-9bb0-cf8e9c2a29a3)

PR: https://github.com/DorisW3/capstone-project/pull/15

**Reply by Kheirah** [reply]
I haven't used `formidable` myself, yet. But from what I can tell, it looks like you have included too much into the callback of `parse`. Within the `parse` function you only validate the data. You don't communicate via `res` just yet.

I don't know if this is going to help, but you can try going in this direction:
```js
export default async function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const data = await parseForm(req);
    const { file } = data.files;

    if (!file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }

    const { newFilename, filepath } = file;
    const result = await cloudinary.v2.uploader.upload(filepath, {
      public_id: newFilename,
      tags: ["some", "example", "tag"],
    });

    return res.status(201).json(result);
  } catch (error) {
    return res.status(500).json({ error: error.message });
  }
}

function parseForm(req) {
  const form = new formidable.IncomingForm();

  return new Promise((resolve, reject) => {
    form.parse(req, (err, fields, files) => {
      if (err) reject(err);
      resolve({ fields, files });
    });
  });
}
```
Mind you, I haven't tested the code. You'll have to do that yourself =)

**Reply by DorisW3** [reply]
This isn't very helpful, because this code is from the handout! Therefore I would expect that it will work!!
